var/spack/repos/builtin/packages/ginkgo/package.py

```diff
@@ -46,12 +47,16 @@ class Ginkgo(CMakePackage, CudaPackage, ROCmPackage):
 
     depends_on("cmake@3.9:", type="build")
     depends_on("cuda@9:", when="+cuda")
+    depends_on("mpi", when="+mpi")
 
-    depends_on("rocthrust", type="build", when="+rocm")
+    depends_on("rocthrust", type="link", when="+rocm")
     depends_on("hipsparse", type="link", when="+rocm")
     depends_on("hipblas", type="link", when="+rocm")
     depends_on("rocrand", type="link", when="+rocm")
     depends_on("hwloc@2.1:", type="link", when="+hwloc")
+    # ROCPRIM is not a direct Ginkgo dependency, but it's a rocthrust dependency
+    # which doesn't play nicely
+    depends_on("rocprim", type="link", when="+rocm")
```

Comment:
rocprim is also header-only. What do you mean when you say it doesn't play nicely?

Reply:
Thanks for a quick answer!

See e.g. https://github.com/spack/spack/pull/33824#issuecomment-1311043479. We internally depend on rocthrust only, but thrust depends on rocprim, and somehow the only fix I can find for this problem is to also depend on rocprim currently.

About the link vs build, link is the closest I can find (we only need the `-I` to be passed as you say) according to https://spack.readthedocs.io/en/latest/packaging_guide.html#dependency-types. But I guess I could leave out the `type` entirely and use the default.

Finally, I tagged you mostly because I had to fix some of the RPATH handling for hiprand, see two comments above: https://github.com/spack/spack/pull/33838#issuecomment-1312464265

Reply:
I would have expected `depends_on("rocthrust", type="build", when="+rocm")` to be sufficient. I suppose I'll need to look into this further to understand the problem.

If the upstream project is using CMake, a call to `find_package(rocthrust REQUIRED)` should search for both rocthrust and its dependencies. A call to `target_link_libraries(<your-lib> PRIVATE roc::rocthrust)` should add the include paths for both rocthrust and rocprim to the library's build flags.

With respect to the rpath stuff... I'm a lot less clear on how that is supposed to work. I'm not sure I can give much useful feedback in a reasonable time frame. The good news is that this is the plan, eventually:

> maybe this package should be split in two so that they rely on standard Spack settings (RPATH, etc).

The bad new is that doesn't help us today.

Reply:
After your comment, I checked and I now see that we simply use targets for everything *but* rocthrust. That explains the problem... Since we've been using the same CMake setup since ROCm 3.5 maybe back then we couldn't rely well on the targets so we went with a simple header approach. That's something we will have to clean up in the future by deprecating older ROCm versions. I'll review the comment.

To explain a bit about the RPATH fix for `rocrand` and `hiprand`, the `$ORIGIN` I add as a patch allows to also set the current directory to the library's RPATH setting so that `libhiprand.so` will finally find the `librocrand.so` installed in the same directory. According to the CI run, this is perfectly fine for the current spack packages, but I do not know if this should be generally pushed into the `rocrand` and `hiprand` repositories.

Also, we seem to be the only package relying on `hiprand` from all of Spack, so only we faced this issue...

Reply:
> I do not know if this should be generally pushed into the rocrand and hiprand repositories

If you're suggesting `upstreaming hiprand_prefer_samedir_rocrand.patch`, I think the answer is "no". As I understand it, the plan is to remove all the RPATH stuff from the default build options in the ROCm libraries. The line you patched that sets the `INSTALL_RPATH` property will probably be removed in a future version. If the packages have been split by that point, then I think everything should 'just work' without any surprises from that point on.

In the meantime, patching the issue in Spack seems like a reasonable approach.

Reply:
Sounds reasonable, thanks!